Add dynamic fee functionality

lok52 opened the pull request.

k1rill-fedoseev commented on CONFIGURATION.md:
```diff
@@ -53,6 +53,11 @@ These environment variables are required for all services.
 | RELAYER_EXPRESS_TRUST_PROXY | If set to `true`, then `trust proxy` express setting will be enabled. Defaults to `false` | boolean |
+| RELAYER_FEE_MANAGER_TYPE | Fee manager that will be used for fee estimations | FeeManagerType |
+| RELAYER_FEE_SCALING_FACTOR | Scaling factor (in percent) used to adjust fee validation. Defaults to 100. | number |
+| RELAYER_PRICE_FEED_TYPE | Price feed type that will be used for rate conversions. | PriceFeedType |
```
I think we need a shortcut feed type for pools with native tokens (e.g. ETH/MATIC/xDAI)

k1rill-fedoseev replied:
Yes

lok52 replied:
Done

k1rill-fedoseev commented on CONFIGURATION.md:
```diff
@@ -53,6 +53,11 @@ These environment variables are required for all services.
 | RELAYER_BLOCKED_COUNTRIES | A list of country codes delimited by spaces | string|
 | RELAYER_EXPRESS_TRUST_PROXY | If set to `true`, then `trust proxy` express setting will be enabled. Defaults to `false` | boolean |
+| RELAYER_FEE_MANAGER_TYPE | Fee manager that will be used for fee estimations | FeeManagerType |
+| RELAYER_FEE_SCALING_FACTOR | Scaling factor (in percent) used to adjust fee validation. Defaults to 100. | number |
```
This way we cannot "scale" user fees, since we always expect them to pay the amount approximately equal to the cost of gas. Let's rename this option to `RELAYER_FEE_MARGIN_FACTOR`, and use second `RELAYER_FEE_SCALING_FACTOR` parameter to proportionally scale results of both: `estimateFee` and `getFees` functions.

So that configuration like:
```
RELAYER_FEE_MARGIN_FACTOR=90
RELAYER_FEE_SCALING_FACTOR=50
```
would mean that the relayer wants to subsidize ~50% of gas costs with 10% margin.

While the configuration like:
```
RELAYER_FEE_MARGIN_FACTOR=90
RELAYER_FEE_SCALING_FACTOR=150
```
would mean that the relayer wants to charge extra 50% on top of its gas costs.

lok52 replied:
Done

EvgenKor commented on zp-relayer/utils/constants.ts:
```diff
@@ -15,6 +15,8 @@ const constants = {
   ZERO_ADDRESS: '0x0000000000000000000000000000000000000000',
   INIT_ROOT: '11469701942666298368112882412133877458305516134926649826543144744382391691533',
+  BASE_CALLDATA_SIZE: 644, // constant calldata size for tx without memo
+  OP_GAS_ORACLE_ADDRESS: '0x420000000000000000000000000000000000000F',
```
Are you sure it's a good idea to define price oracle contract address for a concrete chain as a constant? Maybe it's better to set it as a configuration variable?

lok52 replied:
This is a predefined OP oracle contract deployed at genesis block, so it is not supposed to change in the future. I don't think we should make it a variable parameter, as it will "always" be the same.

k1rill-fedoseev replied:
Will this address be the same for other OP-stack based chains, e.g. Base?

lok52 replied:
Just checked it on goerli testnet, and looks like they use the same address for gas price oracle. The difference, is that it is deployed using a proxy, but this shouldn't matter.
https://goerli.basescan.org/address/0x420000000000000000000000000000000000000f#code

lok52 replied:
>You made separated class and constant for Optimism only. It doesn't look like universal solution

I don't think there is a universal solution for all chains even with a similar fee charging scheme. E.g. for Arbitrum (also an optimistic rollup) we will need a separate class with different logic because they use different contracts with different ABIs . For zkSync the difference will be more significant.

https://developer.arbitrum.io/arbos/l1-pricing

EvgenKor commented:
1. All relayer endpoints related to the token amounts (limits, old fee) return token amount in the pool resolution (gwei). Why do you decide to return wei in the current `/fee/v2` endpoint?
2. Maybe it's better to return `oneByteFee` field in any case to unify response format? E.g. for Polygon pool we can return 
```
{
    "fee": "387049398527485754", // ~0.38$
    "oneByteFee": "0"
}
```

lok52 commented:
> 1. All relayer endpoints related to the token amounts (limits, old fee) return token amount in the pool resolution (gwei). Why do you decide to return wei in the current `/fee/v2` endpoint?
> 2. Maybe it's better to return `oneByteFee` field in any case to unify response format? E.g. for Polygon pool we can return
> 
> ```
> {
>     "fee": "387049398527485754", // ~0.38$
>     "oneByteFee": "0"
> }
> ```

1. Agree, I can adjust the values according to current pool's denominator to simplify client's work
2. Yeah, we can unify the response format for now. But, the problem is that for other pools/chains the fee structure can be different and include other parameters required to calculate the total fee (and the formulas to calculate the fee could be different), so maintaining a unified structure in this case could be difficult.

lok52 commented:
Updates:
* Refactored fee structures
* Added `NativePriceFeed`
* Add fee denomination for `fee/v2` endpoint
* Changed the behaviour of a scaling factor and added a margin factor parameter

akolotov commented on zp-relayer/services/fee/OptimismFeeManager.ts:
```diff
@@ -0,0 +1,81 @@
+
+  async _estimateFee({ gasLimit, data }: IFeeEstimateParams) {
+    const l2Fee = await this.estimateExecutionFee(gasLimit)
+    const l1Fee = await contractCallRetry(this.oracle, 'getL1Fee', [data]).then(toBN)
```
should `data` here be passed together with other transaction's data or their mocks (except, the transaction signature)?

lok52 replied:
Yes, will update it soon

lok52 replied:
I added a constant to account for RLP encoded tx overhead. While it may be not 100% accurate and may vary depending on `chainId` or `gasPrice` length, these fluctuations are not significant.

lok52 commented:
Changes:
* For client side calls OP fee manager now returns `baseFee` as only executionFee
* OP fee manager now accounts for raw RLP encoded tx data overhead

k1rill-fedoseev commented on zp-relayer/endpoints.ts:
```diff
@@ -205,14 +206,26 @@ function relayerInfo(req: Request, res: Response) {
 }
 
-function getFee(req: Request, res: Response) {
+// NOTE: Will be deprecated
+function getStaticFee(req: Request, res: Response) {
```
I think static fees still might be useful for some of our future deployments, e.g. having a static 0.01$ fee for GC deployments might be a good idea.

Can we support all 3 schemes: static, dynamic, dynamic-optimism under the same `/fee` endpoint? We can add a fee manager type field in the JSON response if needed.

lok52 replied:
Good point about GC. I think we can have just one endpoint configurable via some env parameter.

lok52 replied:
I think so

lok52 replied:
Done

k1rill-fedoseev commented on zp-relayer/services/fee/OptimismFeeManager.ts:
```diff
@@ -0,0 +1,109 @@
+    return new FeeEstimate(fee)
+  }
+
+  async _getFees({ gasLimit }: IGetFeesParams): Promise<OptimismUserFeeOptions> {
```
Shouldn't we use `overhead` and `MOCK_CALLDATA` here, so that the UI logic can be abstracted from those?

In the current code, it seems that we will need to repeat similar logic to `_estimateFee ` on the UI side as well.

EvgenKor replied:
I don't want to implement `overhead` calculating logic to the client library because it will not so flexible. From the my point of view the client should exactly know\estimate **_total_** calldata size which will be sent to the relayer. And the client should not think which calldata fields considered to be overhead (with the old logic we should consider deposit transaction's signature to be overhead as well). Moreover the the JS-library doesn't operate with calldata components (it's rust/wasm library job) so we should use length hardcoding anyway here. So current implementation is more universal and convient for the client side

lok52 commented:
Updates:
* Removed `/fee/v2` endpoint. Old `/fee` endpoint remains valid
* Added `StaticFeeManager`. You can configure it by setting `RELAYER_FEE` to pool's token amount in wei (previously it was in gwei) and setting `RELAYER_FEE_MANAGER_TYPE` to `static`

k1rill-fedoseev commented on zp-relayer/services/fee/index.ts:
```diff
@@ -0,0 +1,10 @@
+
+export enum FeeManagerType {
+  Static = 'static',
+  Dynamic = 'Dynamic',
```
```suggestion
  Dynamic = 'dynamic',
```

k1rill-fedoseev commented on zp-relayer/workers/poolTxWorker.ts:
```diff
@@ -133,7 +144,14 @@ export async function createPoolTxWorker({ redis, mutex, txManager, validateTx,
+
+        const requiredFee = await feeManager.estimateFee({
+          gasLimit: config.relayerGasLimit,
+          extraData: tx.rawMemo + (tx.depositSignature || ''),
```
What about other mandatory parts of memo, like withdrawal addresses or permit expiry timestamps? Shouldn't we count them here as well?
